### PR TITLE
bug: show switch authenticator link only when needed

### PR DESCRIPTION
**Pull request by mauriciocastillosilva-okta**
## Description:

- Show the link only if there are more authenticators to choose from. If there is only 1 authenticator, there is no point of showing it
- Context: API will now return the link all the time, so the logic to hide or show it is on the SIW

Resolves: OKTA-399497

## PR Checklist

- [x] Have you verified the basic functionality for this change?
- [x] Added unit tests?
- [x] Added e2e tests
- [ ] Did you follow our [Security Best Practices](https://oktawiki.atlassian.net/wiki/display/eng/Security+Best+practices)?

### Screenshot/Video:

For more context, these are the links:
| Set up | Verify |
| --- | --- |
| SIW-setup authenticator | SIW-verify authenticator |

### Issue:

- [OKTA-399497](https://oktainc.atlassian.net/browse/OKTA-399497)

### CODE REVIEW

File: src/v2/ion/ui-schema/ion-object-handler.js [hunk 1]
```diff
@@ -18,7 +18,7 @@ import { AUTHENTICATOR_KEY } from '../RemediationConstants';
  * @param {AuthenticatorOption[]} options
  * @param {( AuthenticatorEnrollment[] || Authenticator[] )} authenticators
  */
-const createOVOptions = (options = []) => {
+export const createOVOptions = (options = []) => {
```
Comment (anipendakur-okta) [inline src/v2/ion/ui-schema/ion-object-handler.js:21]: Since you're changing this anyway, can we change it to a function instead of a const. -
```
export function createOVOptions(options = []) {
```

Comment (anipendakur-okta) [reply src/v2/ion/ui-schema/ion-object-handler.js:21]: It might be in our best interest to change other exports here too. There's no need for them to be consts as they all are functions. And exporting individual functions supports tree-shaking better (if we were to tree-shake soon.)

File: src/v2/models/AppState.js [hunk 2]
```diff
@@ -77,6 +78,19 @@ export default Model.extend({
...
   },
 
+  hasMoreThanOneAuthenticatorOption(formName) {
+    const form = this.hasRemediationObject(formName);
+    if (form) {
+      const authenticator = form.value.find((value) => value.name === 'authenticator');
+      let authenticatorOptions = authenticator && authenticator.options;
+      // OV is a special case, so process OV options
+      authenticatorOptions = [...authenticatorOptions]; //clone it since we are changing it for OV
+      createOVOptions(authenticatorOptions);
```
Comment (anipendakur-okta) [inline src/v2/models/AppState.js:88]: How about some more ES6 love?
```
let authenticatorOptions = authenticator?.options || [];
authenticatorOptions = [...authenticatorOptions];
```

OR

```
const authenticatorOptions = [...(authenticator?.options || [])];
```

File: src/v2/models/AppState.js [hunk 3]
```diff
@@ -77,6 +78,19 @@ export default Model.extend({
     return this.get('idx').neededToProceed.find((remediation) => remediation.name === formName);
   },
 
+  hasMoreThanOneAuthenticatorOption(formName) {
```
Comment (jmelberg-okta) [inline src/v2/models/AppState.js:81]: name-nit: There may be future cases where we'd like to get the total number of authenticator options. In the spirit of generic function signatures, how about renaming this and changing the return type to an array?

```suggestion
  getAuthenticatorOptions(formName) {
```

File: src/v2/models/AppState.js [hunk 4]
```diff
@@ -77,6 +78,19 @@ export default Model.extend({
...
+    const form = this.hasRemediationObject(formName);
+    if (form) {
```
Comment (jmelberg-okta) [inline src/v2/models/AppState.js:83]: nit: To avoid a large conditional block, consider returning when the following isn't met:

```js
if (!form || !form.value.find(value => value.name === 'authenticator')) {
  // return
}
```

File: src/v2/view-builder/utils/LinksUtil.js [hunk 5]
```diff
@@ -4,7 +4,8 @@ import { ACTIONS, FORMS as RemediationForms } from '../../ion/RemediationConstan
 const { ENROLLED_PASSWORD_RECOVERY_LINK, ORG_PASSWORD_RECOVERY_LINK } = ACTIONS;
 
 const getSwitchAuthenticatorLink = (appState) => {
-  if (appState.hasRemediationObject(RemediationForms.SELECT_AUTHENTICATOR_AUTHENTICATE)) {
+  if (appState.hasRemediationObject(RemediationForms.SELECT_AUTHENTICATOR_AUTHENTICATE) &&
+    appState.hasMoreThanOneAuthenticatorOption(RemediationForms.SELECT_AUTHENTICATOR_AUTHENTICATE)) {
```
Comment (jmelberg-okta) [inline src/v2/view-builder/utils/LinksUtil.js:8]: Since the new method `hasMoreThanOneAuthenticatorOption` checks if this form name exists in the remediation object, we should be able to condense this here and below to the latter part of the conditional.

File: test/testcafe/spec/ChallengeAuthenticatorDuo_spec.js [hunk 6]
```diff
@@ -53,11 +53,7 @@ test
   .requestHooks(mock)('renders an iframe for Duo', async t => {
     const challengeDuoPage = await setup(t);
 
-    const { log } = await t.getBrowserConsoleMessages();
-    await t.expect(log.length).eql(3);
-    await t.expect(log[0]).eql('===== playground widget ready event received =====');
-    await t.expect(log[1]).eql('===== playground widget afterRender event received =====');
-    await t.expect(JSON.parse(log[2])).eql({
```
Comment (jmelberg-okta) [inline test/testcafe/spec/ChallengeAuthenticatorDuo_spec.js:60]: 👏🏼

File: test/testcafe/spec/ChallengeAuthenticatorEmail_spec.js [hunk 7]
```diff
@@ -143,6 +143,11 @@ test
     await t.expect(pageTitle).contains('Verify with your email');
     await t.expect(challengeEmailPageObject.getFormSubtitle())
       .contains('Check your email for a verification message. Click the verification button in your email or enter the code below to continue.');
+
+    // Verify links (switch authenticator link not present since there are no other authenticators available)
+    await t.expect(await challengeEmailPageObject.switchAuthenticatorLinkExists()).notOk();
```
Comment (jmelberg-okta) [inline test/testcafe/spec/ChallengeAuthenticatorEmail_spec.js:148]: Probably OK - but if we modify the mocks at any time we may lose out on this test coverage if responses no longer have a single authenticator present.

For future safety - can we modify the `validOTPmock` above to remove **all but one** authenticator as a separate test?

File: test/unit/spec/v2/view-builder/utils/LinksUtil_spec.js [hunk 8]
```diff
@@ -0,0 +1,43 @@
+import AppState from 'v2/models/AppState';
+import { FORMS } from 'v2/ion/RemediationConstants';
+import { getSwitchAuthenticatorLink } from 'v2/view-builder/utils/LinksUtil';
+
+describe('v2/utils/LinksUtil', function() {
+  const mockAppState = (remediationFormName, hasMoreThanOneAuthenticator) => {
+    const appState = new AppState();
+    spyOn(appState, 'hasRemediationObject').and.callFake(formName => formName === remediationFormName);
+    spyOn(appState, 'hasMoreThanOneAuthenticatorOption').and.returnValue(hasMoreThanOneAuthenticator);
```
Comment (jmelberg-okta) [inline test/unit/spec/v2/view-builder/utils/LinksUtil_spec.js:9]: Let's make sure we use `jest` spies instead of `jasmine`:

```suggestion
    jest.spyOn(appState, 'hasRemediationObject').mockImplementation(formName => formName === remediationFormName);
    jest.spyOn(appState, 'hasMoreThanOneAuthenticatorOption').mockReturnValue(hasMoreThanOneAuthenticator);
```

**Review by jmelberg-okta — COMMENTED** [review]
Great work! Couple of suggestions.

File: test/unit/spec/v2/view-builder/utils/LinksUtil_spec.js (hunk 8 above)
Comment (TianGan-okta) [reply test/unit/spec/v2/view-builder/utils/LinksUtil_spec.js:9]: Bye jasmine

**Comment by gowthamidommety-okta** [comment]
Still reviewing.. question I have is, what will this look like with `ChallengeAuthenticatorDataOktaVerifyView`? This is the view to render OV options when it is the only authenticator

File: src/v2/models/AppState.js (hunk 3 above)
Comment (mauriciocastillosilva-okta) [reply src/v2/models/AppState.js:81]: We don't have the use case for this yet, so I would prefer to keep it as returning a boolean. Once we need this, it will be very easy to refactor this function to support both `hasMoreThanOneAuthenticatorOption` and `getAuthenticatorOptions `

File: test/testcafe/spec/ChallengeAuthenticatorEmail_spec.js (hunk 7 above)
Comment (mauriciocastillosilva-okta) [reply test/testcafe/spec/ChallengeAuthenticatorEmail_spec.js:148]: I don't think we change mocks that often, so I was thinking that we can rely on that mock. If we change it in the future, this test will fail, and I am hoping at that point the developer making the change will do whatever needed to keep the test coverage good. 
PS All the tests I am changing are relying on the mocks as they are today. 

Does this sound good? I can add the test too, but I think we are covered with what we have

**Comment by mauriciocastillosilva-okta** [comment]
@gowthamidommety-okta if OV is the only authenticator we look at the methods that are available. This is what `createOVOptions` function does. So anytime, OV been the single authenticator or not, OV will be counted depending how many methods are available

File: src/v2/ion/ui-schema/ion-object-handler.js [hunk 9]
```diff
@@ -147,6 +147,9 @@ const createUiSchemaForObject = (ionFormField, remediationForm, transformedResp,
...
 
   return uiSchema;
-};
+}
 
-export default createUiSchemaForObject;
+export {
+  createUiSchemaForObject as default,
+  createOVOptions
+};
```
Comment (anipendakur-okta) [inline src/v2/ion/ui-schema/ion-object-handler.js:155]: Exporting an object will actually import the whole object into the bundle as compared to individual exports. In order to support tree-shaking we would need something like
```
export function createOVOptions() {}
.
.
.
export default createUiSchemaForObject;
```

File: src/v2/ion/ui-schema/ion-object-handler.js [hunk 10]
```diff
@@ -18,7 +18,7 @@ import { AUTHENTICATOR_KEY } from '../RemediationConstants';
  * @param {AuthenticatorOption[]} options
  * @param {( AuthenticatorEnrollment[] || Authenticator[] )} authenticators
  */
-const createOVOptions = (options = []) => {
+function createOVOptions(options = []) {
```
Comment (pradeepdewda-okta) [inline src/v2/ion/ui-schema/ion-object-handler.js:21]: why this change ??

File: src/v2/ion/ui-schema/ion-object-handler.js (hunk 9 above)
Comment (pradeepdewda-okta) [reply src/v2/ion/ui-schema/ion-object-handler.js:155]: `export default` make sense..

File: src/v2/models/AppState.js (hunk 3 above)
Comment (pradeepdewda-okta) [reply src/v2/models/AppState.js:81]: we can make this function more generic right away..

hasRequiredAuthenticators(formName, count) {
    .
    .
   return authenticatorOptions.length > count;
}

File: test/testcafe/spec/ChallengeAuthenticatorEmail_spec.js [hunk 11]
```diff
@@ -143,6 +143,11 @@ test
...
+    await t.expect(await challengeEmailPageObject.signoutLinkExists()).ok();
+    await t.expect(challengeEmailPageObject.getSignoutLinkText()).eql('Back to sign in');
```
Comment (pradeepdewda-okta) [inline test/testcafe/spec/ChallengeAuthenticatorEmail_spec.js:150]: There could two types of `Back to sign in` links..  `js-cancel` and `js-goback` . If `cancel` object is not present `js-goback` will be rendered.. just fyi

File: test/testcafe/spec/EnrollAuthenticatorPhoneView_spec.js [hunk 12]
```diff
@@ -83,6 +85,9 @@ test
     await t.expect(pageSubtitle).contains('your phone');
     await t.expect(pageSubtitle).contains('Enter the code below to verify.');
 
+    // Verify links
```
Comment (pradeepdewda-okta) [inline test/testcafe/spec/EnrollAuthenticatorPhoneView_spec.js:88]: dups ?

File: test/unit/spec/v2/models/AppState_spec.js [hunk 13]
```diff
@@ -180,4 +185,130 @@ describe('v2/models/AppState', function() {
     });
 
   });
+
+  describe('hasRemediationObject', () => {
+    it('returns null if there form is not present', (done) => {
```
Comment (pradeepdewda-okta) [inline test/unit/spec/v2/models/AppState_spec.js:190]: `if there form` ??

File: test/unit/spec/v2/view-builder/utils/LinksUtil_spec.js [hunk 14]
```diff
@@ -0,0 +1,37 @@
+import AppState from 'v2/models/AppState';
```
Comment (pradeepdewda-okta) [inline test/unit/spec/v2/view-builder/utils/LinksUtil_spec.js:1]: 🥇

File: src/v2/ion/ui-schema/ion-object-handler.js (hunk 10 above)
Comment (mauriciocastillosilva-okta) [reply src/v2/ion/ui-schema/ion-object-handler.js:21]: if you scroll up the comments in the PR, Ani pointed out why https://github.com/okta/okta-signin-widget/pull/2006#discussion_r645057750

File: src/v2/models/AppState.js (hunk 3 above)
Comment (mauriciocastillosilva-okta) [reply src/v2/models/AppState.js:81]: ok, I got a couple of comments on this function, so I will refactor it. I prefer Jordan's proposal, which is to return the options objects and let the consumer of the function decide what to do with them (either look at the length or iterate through them). Will make the change

File: src/v2/models/AppState.js [hunk 15]
```diff
@@ -77,6 +78,19 @@ export default Model.extend({
     return this.get('idx').neededToProceed.find((remediation) => remediation.name === formName);
   },
 
+  getRemediationAuthenticationOptions(formName) {
+    const form = this.hasRemediationObject(formName);
+    if (!form) {
+      return false;
```
Comment (jmelberg-okta) [inline src/v2/models/AppState.js:84]: Now that this isn't returning a `boolean`, let's return an empty array here to ensure a uniform type.

**Review by anipendakur-okta — APPROVED** [review]
LGTM

File: test/testcafe/spec/ChallengeOktaVerifyTotp_spec.js [hunk 16]
```diff
@@ -52,6 +52,12 @@ test
     await t.expect(await challengeOktaVerifyTOTPPageObject.getTotpLabel())
       .contains('Enter code from Okta Verify app');
 
+    // Verify links
+    await t.expect(await challengeOktaVerifyTOTPPageObject.switchAuthenticatorLinkExists()).ok();
+    await t.expect(challengeOktaVerifyTOTPPageObject.getSwitchAuthenticatorLinkText()).eql('Verify with something else');
+    await t.expect(await challengeOktaVerifyTOTPPageObject.signoutLinkExists()).ok();
+    await t.expect(challengeOktaVerifyTOTPPageObject.getSignoutLinkText()).eql('Back to sign in');
```
Comment (jmelberg-okta) [inline test/testcafe/spec/ChallengeOktaVerifyTotp_spec.js:59]: Looking at this again, I think we have an opportunity to condense these tests. I created a follow-up for us to look into breaking out the footer checking/content logic out to a shared assertion ([OKTA-402002](https://oktainc.atlassian.net/browse/OKTA-402002)).

File: test/testcafe/spec/EnrollAuthenticatorGoogleAuthenticator_spec.js [hunk 17]
```diff
@@ -117,6 +122,10 @@ test
     await t.expect(enrollGoogleAuthenticatorPageObject.getSharedSecret()).eql('ZR74DHZTG43NBULV');
     await enrollGoogleAuthenticatorPageObject.goToNextPage();
 
+    // Verify links (switch authenticator link not present since there are no other authenticators available)
```
Comment (jmelberg-okta) [inline test/testcafe/spec/EnrollAuthenticatorGoogleAuthenticator_spec.js:125]: These comments are extremely helpful as reviewers walk through the code. Thank you!

File: test/unit/spec/v2/view-builder/utils/LinksUtil_spec.js [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+import { FORMS } from 'v2/ion/RemediationConstants';
+import { getSwitchAuthenticatorLink } from 'v2/view-builder/utils/LinksUtil';
+
+describe('v2/utils/LinksUtil', function() {
+  const mockAppState = (remediationFormName, hasMoreThanOneAuthenticator) => {
+    const appState = new AppState();
+    spyOn(appState, 'getRemediationAuthenticationOptions').and.callFake(formName => {
```
Comment (jmelberg-okta) [inline test/unit/spec/v2/view-builder/utils/LinksUtil_spec.js:8]: Similar to the other changes, let's make sure there are no `spyOn` methods being called that aren't methods on `jest`. This helps us eventually migrate to future major versions of Jest, [which plans](https://jestjs.io/blog/2020/05/05/jest-26#a-new-way-to-consume-jest---jestglobals) to eventually drop these as a part of their global objects.

```suggestion
    jest.spyOn(appState, 'getRemediationAuthenticationOptions').mockImplementation(formName => {
```

**Review by jmelberg-okta — APPROVED** [review]
Solid work! 💪🏼 

Couple of minor comments + changes.

File: test/testcafe/spec/ChallengeOktaVerifyTotp_spec.js [hunk 19]
```diff
@@ -52,6 +52,12 @@ test
     await t.expect(await challengeOktaVerifyTOTPPageObject.getTotpLabel())
       .contains('Enter code from Okta Verify app');
 
+    // Verify links
+    await t.expect(await challengeOktaVerifyTOTPPageObject.switchAuthenticatorLinkExists()).ok();
+    await t.expect(challengeOktaVerifyTOTPPageObject.getSwitchAuthenticatorLinkText()).eql('Verify with something else');
```
[the remark this answers is not in the document]
Comment (mauriciocastillosilva-okta) [reply test/testcafe/spec/ChallengeOktaVerifyTotp_spec.js:57]: You know, this is a very good question. I don't know either why we check for both, but I have also seen it in different places so I copied the pattern. My only guess is that we do this in case there is more than one link and one of them is empty? Which is really unlikely... would it be ok if we clean this up in the jira you filed? [OKTA-402002](https://oktainc.atlassian.net/browse/OKTA-402002) (which I know is for the signout link... or I can file another jira too)

**Review by gowthamidommety-okta — APPROVED** [review]
Sorry for the late comment. Everything looks good, but I'm still concerned about OV when it is the only authenticator. Can we add tests here https://github.com/okta/okta-signin-widget/blob/master/test/testcafe/spec/ChallengeAuthenticatorOktaVerify_spec.js#L5 to make sure it works as expected?

File: test/testcafe/spec/ChallengeAuthenticatorOktaVerify_spec.js [hunk 20]
```diff
@@ -128,3 +137,36 @@ test.requestHooks(requestLogger, mockChallengeOVSelectMethod)('should send right
     'stateHandle': '02ciZ1YTWakSanNu8GNNTnTXzhL5hoLzTlR0JewfG3'
   });
 });
+
+test.requestHooks(requestLogger, mockChallengeOVTotpMethod)('should show switch authenticator link only when needed', async t => {
+  const selectAuthenticatorPage = await setup(t);
+  await t.expect(selectAuthenticatorPage.getFormTitle()).eql('Verify it\'s you with an authenticator');
+  await t.expect(selectAuthenticatorPage.getFormSubtitle()).eql('Select from the following options');
+
+  // This view (only OV authenticator, but with 3 methods) is custom
```
Comment (gowthamidommety-okta) [inline test/testcafe/spec/ChallengeAuthenticatorOktaVerify_spec.js:146]: 🙌

**Review by gowthamidommety-okta — APPROVED** [review]
Looks great! Thank you so much for checking 🙏
